### ISSUE TITLE
Installing on Mac OSX

### ISSUE THREAD

**Issue by bahoo** [rprichard/x11-canvas-screencast#1]
Hello,

I can't get past this error message when running `make`:

```
$ qmake
$ make
clang++ -c -pipe -stdlib=libc++ -mmacosx-version-min=10.7 -O2 -arch x86_64 -Wall -W -DQT_NO_DEBUG -DQT_GUI_LIB -DQT_CORE_LIB -DQT_SHARED -I/usr/local/Cellar/qt/4.8.6/mkspecs/unsupported/macx-clang-libc++ -I. -I/usr/local/Cellar/qt/4.8.6/lib/QtCore.framework/Versions/4/Headers -I/usr/local/Cellar/qt/4.8.6/lib/QtCore.framework/Versions/4/Headers -I/usr/local/Cellar/qt/4.8.6/lib/QtGui.framework/Versions/4/Headers -I/usr/local/Cellar/qt/4.8.6/lib/QtGui.framework/Versions/4/Headers -I/usr/local/Cellar/qt/4.8.6/include -I/opt/X11/include -I. -F/usr/local/Cellar/qt/4.8.6/lib -o Cursor.o Cursor.cpp
In file included from Cursor.cpp:12:
./CaptureConfig.h:6:1: warning: 'CaptureConfig' defined as a struct here but previously declared as a class [-Wmismatched-tags]
struct CaptureConfig
^
./Cursor.h:11:1: note: did you mean struct here?
class CaptureConfig;
^~~~~
struct
1 warning generated.
clang++ -headerpad_max_install_names -stdlib=libc++ -mmacosx-version-min=10.7 -arch x86_64 -o screencast main.o MurmurHash3.o Cursor.o CaptureConfig.o X11Bridge.o   -F/usr/local/Cellar/qt/4.8.6/lib -L/usr/local/Cellar/qt/4.8.6/lib -L/opt/X11/lib -lX11 -lXfixes -framework QtGui -L/usr/local/Cellar/qt/4.8.6/lib -F/usr/local/Cellar/qt/4.8.6/lib -framework QtCore 
Undefined symbols for architecture x86_64:
  "QX11Info::display()", referenced from:
      _main in main.o
      screencast::Cursor::Cursor(screencast::CaptureConfig const&) in Cursor.o
      screencast::capsLockEnabled() in X11Bridge.o
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make: *** [screencast] Error 1
$ 
```

I've installed qt via Homebrew using `brew install qt --developer` which purports to "Build and link with developer options", but, no dice.

I see what appears to be the source for QX11Info residing in `/usr/local/Cellar/qt/4.8.6/lib/QtGui.framework/` but I'm not smart enough to connect the dots on where the compiled version might have ended up.

Any ideas? Thank you.

**Comment by rprichard** [comment]
Sorry, I don't know where the compiled version of QX11Info is at.  If it exists, I would have expected it to be linked in automatically as on Linux.  I think your Qt installation has a bug if QX11Info can be included, but the object code isn't linked in.  I tried building x11-canvas-screencast with a copy of OS X Qt 5.1.1, and I got errors about missing x11 and xfixes packages.  It's not clear to me whether OS X Qt has any support for X11.

IIRC, x11-canvas-screencast only uses X11 for two things:
- Detecting whether CAPSLOCK is on or off (i.e. to pause/unpause recording without moving the mouse)
- Capturing the mouse cursor icon.

I wonder if an X11 program on OS X can actually do these two things, and in any case, maybe there's a non-X11 method.

**Comment by rprichard** [comment]
FWIW, I tried removing the use of QX11Info.  It's on the `osx` branch.  I'll test it on OS X, but I'm waiting for `brew install qt --developer` to finish.  (Also, will that brew command give me the X11/Xfixes headers?)

**Comment by bahoo** [comment]
Ooh, thank you for the follow-through here, much appreciated.

I'm a web guy by trade so I'm pretty fish-outta-water in compiling from source. But that `brew` command produced a message that it couldn't find the X11 package, which I resolved by adding the right path to `$PKG_CONFIG_PATH`. It never barked about xfixes.

After some thorough Googling, I found a few links that might be of more help here?

http://unix.stackexchange.com/questions/42334/how-to-fix-link-error-under-cmake-for-class-qx11info-on-mac-os-x
https://lists.macosforge.org/pipermail/macports-tickets/2009-March/026151.html
https://trac.macports.org/ticket/18724
https://github.com/thoughtbot/capybara-webkit/wiki/Installing-Qt-and-compiling-capybara-webkit

**Comment by rprichard** [comment]
I tested my `osx` branch change.  The code _almost_ works.  It starts up XQuartz on the first run.  It notices me toggling CAPSLOCK on and off.  It also successfully uses Xfixes to capture the mouse cursor icon, with correct transparency handling.

It doesn't read the cursor position correctly, though.  It seems to think it's always stuck in the center of the screen.

**Comment by rprichard** [comment]
I spoke too soon.  The code uses `XFixesGetCursorImage` to read both the cursor image and the cursor position/hotspot.  That API isn't really implemented -- it's just returning a stock arrow cursor in the center of the screen, regardless of the real cursor being displayed.

It looks it's easy to do a proper OS X port, though.
- Call `[NSEvent modifierFlags]` to check whether CAPSLOCK is toggled.
- Use `[NSCursor currentSystemCursor]`'s `hotSpot` and `image` methods to get the cursor image.
- Use `[NSEvent mouseLocation]` or `QCursor::pos()` to get the cursor position.

I'll take a crack at it soon-ish.

**Comment by bahoo** [comment]
Awesome. The `osx` branch now compiles & captures fine for me. It ran out of memory doing ./pack_animation but that's my own problem. Thanks again!

**Comment by rprichard** [comment]
I've pushed commits to master that remove the X11 dependency on OS X.  Let me know how it works for you.

**Comment by bahoo** [comment]
Sweet, yes, it's working. Thank you.

For other folks who might be Googling: I did run into segfaults when running `pack_animation.py` but looks like that's an OSX / Python bug in opencv. http://stackoverflow.com/questions/24084391/segmentation-fault-and-crashing-when-trying-to-import-opencv I followed their instructions and also had to replace my shebang with a hard-coded path to Python in my virtualenv. Not great, but it'll do the job for the time being.

Thanks again, Ryan!
